### DISCUSSION TITLE
Doorbell notifications very unreliable

### DISCUSSION THREAD

**Discussion by delingren** [HomeSpan/HomeSpan#960] [Q&A] [answered]
I am playing with the doorbell service. It looks pretty straightforward. I added it to the Home app and the tile says "Not supported". I looked up the FAQ and it says it'll chime the HomePod anyway. I don't have a HomePod. I have an AppleTV. So I supposed it'd push a notification to my iPhone.
However, it turned out that the notification happens very inconsistently. I only get one notification out of 10-20 button pushes. And I can't seem to find a pattern when it works and when it doesn't. I'm a little baffled. Notifications from all my other HomeSpan devices are working perfectly fine (contact sensors, temp sensors, garage door opener).
Has anyone else seen anything like this?

And here's the entirety of my sketch:
```
#include "HomeSpan.h"

struct Doorbell : Service::Doorbell {
  SpanCharacteristic *switchEvent;

  Doorbell(int pin) : Service::Doorbell() {
    switchEvent = new Characteristic::ProgrammableSwitchEvent();
    new SpanButton(pin, 2000, 5, 200, SpanButton::TRIGGER_ON_LOW);
  }

  void button(int pin, int pressType) override {
    switchEvent->setVal(pressType);
    Serial.println("Ding dong!");
  }
};

void setup() {
  Serial.begin(115200);
  homeSpan.begin(Category::ProgrammableSwitches, "Doorbell");

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Doorbell(26);

  homeSpan.autoPoll();
}

void loop() {}
```

**Comment by delingren** [comment]
After playing with it a bit more, I think I have found the pattern. A press immediately following a previous press of the same type (single, double, or long) doesn't trigger a notification. But if the second press is a different type, it does trigger the notification. That's interesting. Is this by design? I.e. does HomeKit intentionally not to send multiple notifications of the same type?

**Reply by HomeSpan** [reply]
What do the log files from HomeSpan show?  Are you seeing it send notifications for each press (set log level=2 in the CLI to see all log info), or is it sending notifications only if you change press-types?   Need to determine if the behavior is because of something in HomeSpan, or rather it is on the HomeKit side.

**Reply by delingren** [reply]
Here's the log for the two consecutive presses. The first one triggered the notification. The second one did not. The log is exactly the same.

Another observation is that, after a couple of minutes, the same type press would trigger the notification again. It almost seems like the switch is stateful and the state is reset after a period of time.

I can't imagine this would be the intended behavior of a doorbell. Each press has to register a notification, right? 

```
21:04:37.810 -> Ding dong!
21:04:37.810 -> 0
21:04:37.810 -> 
21:04:37.810 -> >>>>>>>>>> 192.168.1.167 >>>>>>>>>>
21:04:37.810 -> EVENT/1.0 200 OK
21:04:37.811 -> Content-Type: application/hap+json
21:04:37.811 -> Content-Length: 49
21:04:37.811 -> 
21:04:37.811 -> {"characteristics":[{"iid":4,"value":0,"aid":1}]}
21:04:37.842 -> -------- SENT ENCRYPTED! --------
21:04:41.713 -> Ding dong!
21:04:41.713 -> 0
21:04:41.713 -> 
21:04:41.713 -> >>>>>>>>>> 192.168.1.167 >>>>>>>>>>
21:04:41.713 -> EVENT/1.0 200 OK
21:04:41.713 -> Content-Type: application/hap+json
21:04:41.713 -> Content-Length: 49
21:04:41.713 -> 
21:04:41.713 -> {"characteristics":[{"iid":4,"value":0,"aid":1}]}
21:04:41.713 -> -------- SENT ENCRYPTED! --------
```

**Comment by HomeSpan** [comment, marked as answer]
According to the log file HomeSpan is sending the notification each time (it's the line that says "EVENT").  What HomeKit does with the notification is outside the control of HomeSpan.  I just tested the sketch you have above without any modifications and it works fine - my HomePod chimes whenever the button is pressed, though if I press it a second time while the chime noise is still playing it waits for the chime to stop before retriggered (about 2-second delay).

BUT...what I do see is that I do **not** get a notification **every** time on my iPhone or MAC when the button is pressed.  I get repeated chimes, but the notification only shows up once even if the chime sounds multiple times for multiple presses all in a row.  If I wait a minute or so a notification gets sent again once I press, as you've observed above.  The only thing you were missing was the HomePod to confirm that HomeKit was indeed processing the repeated button presses.

So it looks like everything is working totally fine - what you've discovered is that the iPhone seems to know not to repeat notifications of repeated events that occur in succession, even if the chime does repeat as expected.  Not sure if there are iPhone settings that allow you to modify the timing of when a second notification of the same type will be sent again.  If you find any, please post here in case others are interested.

**Reply by delingren** [reply]
Thanks for the analysis. Sounds like you got the same results on your iPhone and Mac. I have 4 iPhones, an iPad, and 3 Macs (I know, I know, I need to get rid of some of them) and they all behave the same way. I double checked every setting and I didn't see anything relevant. I guess it *is* by design then. I suppose the door bell is mainly meant to notify people at home.

If I ever come across any info on this matter, I'll update this thread.
